### PR TITLE
Manage resource iterators explicitly

### PR DESCRIPTION
**Pull request by andreikoval**
Knowledge about iterators' inherent resources must not leak outside
relevant scope.  Responsibility of managing such resources also must be
explicit, not hidden as an implicit "instanceof" check in the totally
irrelevant part of code.

### CODE REVIEW

**Review by burqen — CHANGES_REQUESTED** [review]
Changes look good! However, we add one extra layer if wrapping to a lot of index calls. I'll start a benchmark run to see how this affect performance. Hold tight.

**Comment by burqen** [comment]
Benchmarks look fine, approved!

**Review by burqen — DISMISSED** [review]

**Comment by andreikoval** [comment]
As discussed offline, benchmark ui was confusing and there's actually a regression. Alternative solution will be implemented.

File: community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableRelationshipDiffSets.java [hunk 1]
```diff
@@ -128,4 +123,26 @@ public void accept( DiffSetsVisitor<T> visitor )
         {
         }
     }
+
+    RelationshipIterator augment( RelationshipIterator source );
+
+    RelationshipIterator augmentWithRemovals( RelationshipIterator source );
```
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableRelationshipDiffSets.java]: This guy is never used

File: community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableRelationshipDiffSets.java [hunk 2]
```diff
@@ -24,13 +24,14 @@
...
-import org.neo4j.collection.primitive.PrimitiveIntIterator;
+import org.neo4j.collection.primitive.PrimitiveLongIterator;
+import org.neo4j.collection.primitive.PrimitiveLongResourceIterator;
 import org.neo4j.kernel.impl.api.store.RelationshipIterator;
 
 /**
  * {@link SuperReadableDiffSets} with added method for filtering added relationships.
  */
-public interface ReadableRelationshipDiffSets<T> extends SuperReadableDiffSets<T,RelationshipIterator, RelationshipIterator>
+public interface ReadableRelationshipDiffSets<T> extends SuperReadableDiffSets<T>
```
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableRelationshipDiffSets.java:34]: Can we place `Empty` implementation at the bottom and the interesting stuff (the methods) above?

**Review by burqen — APPROVED** [review]
This is so much cleaner, I really like this! Nice 👍
I'll approve this now, but please have a look at the small readability comments before merging.

**Comment by burqen** [comment]
The test failure is not related to this PR. A rebase should fix it.

**Comment by andreikoval** [comment]
Cool, thanks! I'll fix those readability issues.

**Pull request merged**